### PR TITLE
bugfix: global transaction hook repeat execute

### LINKED ISSUE #5595: Repeatedly executed hook in TransactionalTemplate
- [ ] I have searched the [issues](https://github.com/seata/seata/issues) of this repository and believe that this is not a duplicate.

### Ⅰ. Issue Description
在 https://github.com/seata/seata/pull/5594 里修复了日志重复打印的问题，但hook重复执行依然存在。以begin为例子，每次切面进来时会先执行hook再执行tx.begin()。
image
在tx.begin()里面会有 判断role != GlobalTransactionRole.Launcher并直接返回，但before和after的hook依然执行了，这个不太符合事务传播的逻辑，因为根本就没有新开启事务
image

同理，在commit和rollback都有这种问题。大家可以讨论一下是不是要改成只执行一次。
另外我有考虑过，如果把一些关键参数给到hook，用户就能自行决定，不过这样做有点重，可能没有必要。
如果确认需要改成执行一次的话，我可以fix掉

### Ⅱ. Describe what happened

  If there is an exception, please attach the exception trace:

```
Just paste your stack trace here!
```

### Ⅳ. How to reproduce it (as minimally and precisely as possible)

1. xxx
2. xxx
3. xxx

Minimal yet complete reproducer code (or URL to code):

### Ⅵ. Environment:

- JDK version(e.g. `java -version`):
- Seata client/server version:  
- Database version:
- OS(e.g. `uname -a`):
- Others:

### PR DESCRIPTION
**Pull request by jsbxyyx**
- [ ] I have registered the PR [changes](../changes).

### Ⅱ. Does this pull request fix one issue?

fixes #5595

### CODE REVIEW

**Review by slievrly — APPROVED** [review]
LGTM
add change

**Review by slievrly — COMMENTED** [review]
@jsbxyyx Pls. check the UT.

File: tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java [hunk 1]
```diff
@@ -286,82 +286,96 @@ private void rollbackTransaction(GlobalTransaction tx, Throwable originalExcepti
...
     }
 
-    private void triggerAfterCompletion() {
-        for (TransactionHook hook : getCurrentHooks()) {
-            try {
-                hook.afterCompletion();
-            } catch (Exception e) {
-                LOGGER.error("Failed execute afterCompletion in hook {}", e.getMessage(), e);
+    private void triggerAfterCompletion(GlobalTransaction tx) {
+        if (tx == null || tx.getGlobalTransactionRole() == GlobalTransactionRole.Launcher) {
```
Comment (leizhiyuan) [inline tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java:372]: which scene  tx can be null ? it seems tx can not null from the invoke context

Comment (jsbxyyx) [reply tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java:372]: 这里不会为null，从事务模型上来说，事务开启失败了，是否需要调用completion

Comment (leizhiyuan) [reply tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java:372]: 我看其他的hook判断条件里面，都认为tx不会为null，所以都没有这个tx==null的判断，但是triggerAfterCompletion 这个需要有这个特殊处理。后面修改的人可能会疑惑

如果认为这里tx 也不会为null，是不是可以吧tx==null的条件去掉，和其他修改的地方保持一致？

Comment (jsbxyyx) [reply tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java:372]: 这里的判断是和saga保持一致的

**Review by funky-eyes — COMMENTED** [review]
The hook is called by TransactionalTemplate, and the Participant's judgment is inside the class defaultGlobaltransaction, so I'm wondering again if I should underlay the hook call into defaultGlobaltransaction? Or the other way around, make defaultGlobaltransaction's behavior more pure, and all the prevention of duplicate calls and permission control is handled by TransactionalTemplate?
------------
钩子是TransactionalTemplate调的，而Participant的判断是在defaultGlobaltransaction这个类里面的，所以我再想是否应该将hook的调用下层到defaultGlobaltransaction中？或者反过来，让defaultGlobaltransaction的行为更加单纯，所有的防止重复调用，权限控制都由TransactionalTemplate来处理？

**Comment by jsbxyyx** [comment]
> The hook is called by TransactionalTemplate, and the Participant's judgment is inside the class defaultGlobaltransaction, so I'm wondering again if I should underlay the hook call into defaultGlobaltransaction? Or the other way around, make defaultGlobaltransaction's behavior more pure, and all the prevention of duplicate calls and permission control is handled by TransactionalTemplate?
> 
> 钩子是TransactionalTemplate调的，而Participant的判断是在defaultGlobaltransaction这个类里面的，所以我再想是否应该将hook的调用下层到defaultGlobaltransaction中？或者反过来，让defaultGlobaltransaction的行为更加单纯，所有的防止重复调用，权限控制都由TransactionalTemplate来处理？

角色判断修改为提前处理

**Review by funky-eyes — APPROVED** [review]
LGTM

**Comment by funky-eyes** [comment]
image
Unit tests need to be fixed before merging can take place

**Pull request merged**